File: src/runner.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufReader};
use std::process::ExitCode;

use crate::{counter, flags};

pub fn run(args: Vec<String>) -> ExitCode {
    // Parse flags and filepaths, printing any errors encountered
    let (flags, filepaths) = match flags::parse_flags_and_filenames(args) {
        Ok(result) => result,
        Err(e) => {
            eprintln!("{e}");
            return ExitCode::FAILURE;
        }
    };

    if filepaths.is_empty() {
        process_stdin(&flags)
    } else {
        process_filepaths(flags, filepaths)
    }
}

fn process_stdin(flags: &HashSet<flags::Flag>) -> ExitCode {
    let stdin = io::stdin();
    let counts = counter::count(&mut stdin.lock(), flags);
    counter::print_counts(flags, &counts);
    println!();
    ExitCode::SUCCESS
}

fn process_filepaths(flags: HashSet<flags::Flag>, filepaths: Vec<String>) -> ExitCode {
    let mut status = ExitCode::SUCCESS;

    for filepath in &filepaths {
        process_filepath(filepath, &flags, &mut status);
    }

    status
}

fn process_filepath(filepath: &str, flags: &HashSet<flags::Flag>, status: &mut ExitCode) {
    match File::open(filepath) {
        Ok(file) => {
            let counts = counter::count(&mut BufReader::new(file), flags);
            counter::print_counts(flags, &counts);
            println!(" {filepath}");
        }
        Err(_) => {
            eprintln!("ccwc: {filepath}: No such file or directory");
            *status = ExitCode::FAILURE;
        }
    }
}

File: zig/cc-wc-core/src/runner.zig
const std = @import("std");
const arguments = @import("cc-wc-core-internal").arguments;
const counter = @import("cc-wc-core-internal").counter;
const FlagSet = @import("cc-wc-core-internal").flags.FlagSet;

pub fn run(allocator: std.mem.Allocator, args: []const []const u8) u8 {
    const parsed = arguments.parseFlagsAndFilenames(args) catch return 1;

    if (parsed.files.len == 0) {
        return processStdin(allocator, parsed.flags);
    }
    return processFiles(allocator, parsed.flags, parsed.files);
}

fn processStdin(allocator: std.mem.Allocator, flags: FlagSet) u8 {
    const counts = counter.count(allocator, std.fs.File.stdin().deprecatedReader(), flags) catch return 1;
    counter.printCounts(flags, counts);
    std.fs.File.stdout().deprecatedWriter().print("\n", .{}) catch {};
    return 0;
}

fn processFiles(allocator: std.mem.Allocator, flags: FlagSet, filepaths: []const []const u8) u8 {
    var status: u8 = 0;
    for (filepaths) |filepath| {
        processFile(allocator, flags, filepath, &status);
    }
    return status;
}

fn processFile(allocator: std.mem.Allocator, flags: FlagSet, filepath: []const u8, status: *u8) void {
    const file = std.fs.cwd().openFile(filepath, .{}) catch {
        std.fs.File.stderr().deprecatedWriter().print("ccwc: {s}: No such file or directory\n", .{filepath}) catch {};
        status.* = 1;
        return;
    };
    defer file.close();

    const counts = counter.count(allocator, file.deprecatedReader(), flags) catch {
        status.* = 1;
        return;
    };
    counter.printCounts(flags, counts);
    std.fs.File.stdout().deprecatedWriter().print(" {s}\n", .{filepath}) catch {};
}
